sql/src/main/java/io/crate/operation/projectors/IndexWriterCountBatchIterator.java

```diff
@@ -0,0 +1,187 @@
+    private final BatchIterator source;
+    private CompletableFuture<BitSet> loading;
+    private Row currentRow = OFF_ROW;
+    private BitSet result;
+
+    private IndexWriterCountBatchIterator(BatchIterator source,
+                                          Supplier<String> indexNameResolver,
+                                          Input<BytesRef> sourceInput,
+                                          Iterable<? extends CollectExpression<Row, ?>> collectExpressions,
+                                          RowShardResolver rowShardResolver,
+                                          BulkShardProcessor bulkShardProcessor) {
+        this.indexNameResolver = indexNameResolver;
+        this.collectExpressions = collectExpressions;
+        this.sourceInput = sourceInput;
+        this.rowShardResolver = rowShardResolver;
+        this.bulkShardProcessor = bulkShardProcessor;
+        this.source = source;
+    }
+
+    public static BatchIterator newInstance(BatchIterator source, Supplier<String> indexNameResolver,
+                                            Input<BytesRef> sourceInput,
+                                            Iterable<? extends CollectExpression<Row, ?>> collectExpressions,
+                                            RowShardResolver rowShardResolver,
+                                            BulkShardProcessor bulkShardProcessor) {
+        IndexWriterCountBatchIterator delegate = new IndexWriterCountBatchIterator(source, indexNameResolver,
+            sourceInput, collectExpressions, rowShardResolver, bulkShardProcessor);
+        return new CloseAssertingBatchIterator(delegate);
+    }
+
+    @Override
+    public void moveToStart() {
+        raiseIfLoading();
+        currentRow = OFF_ROW;
+        result = null;
+    }
+
+    @Override
+    public boolean moveNext() {
+        if (loading == null) {
+            return false;
+        }
+        raiseIfLoading();
+
+        if (currentRow.equals(OFF_ROW) && result == null) {
+            result = loading.join();
+            long rowCount = result == null ? 0 : result.cardinality();
+            currentRow = new Row1(rowCount);
+            return true;
+        }
+
+        currentRow = OFF_ROW;
+        return false;
+    }
+
+    @Override
+    public Row currentRow() {
+        raiseIfLoading();
+        return currentRow;
+    }
+
+    @Override
+    public void close() {
+        source.close();
+        bulkShardProcessor.close();
+    }
+
+    @Override
+    public CompletionStage<?> loadNextBatch() {
+        if (isLoading()) {
+            return CompletableFutures.failedFuture(new IllegalStateException("Iterator is already loading"));
+        }
+
+        if (allLoaded()) {
+            return CompletableFutures.failedFuture(new IllegalStateException("All batches already loaded"));
+        }
+        loading = new CompletableFuture();
+        bulkShardProcessor.result().whenComplete((r, t) -> {
```

Comment:
I initially thought about that too, but I don't think it's a good idea to expose it to the client code. 

There's also a couple of shortcuts where we fail the loading future earlier, if the source yields an exception while consuming it for example.